Navigation: Implement new design for sub-menus

retrofox opened the pull request:
## Description
This PR improves the sub-menus for the Navigation block. Significant changes:

* Apply position absolute to sub-menus, removing the block movements when a sub-menu is opened/closed, making the UX less stressful for the user.
* Propagate text and background colors to the sub-menus.
* Apply Light/Dark styles only when it doesn't have a custom background color.
* ~Apply the border-color getting its value from text color ([to be defined/under discussion](https://github.com/WordPress/gutenberg/issues/18310#issuecomment-576441985))~

Fixes https://github.com/WordPress/gutenberg/issues/18310

### Design

The [final design to implement on this PR was defined](https://github.com/WordPress/gutenberg/issues/18310#issuecomment-574737902) in #18310:

<img src="https://user-images.githubusercontent.com/191598/72371352-377a1080-36d2-11ea-9703-ddaff11d9663.gif" width="600px" />

### About technical implementation

We need to keep in mind that this PR propagates the color values from the Navigation block to each Navigation Item. It simplifies applying the styles via CSS, and get ready the code to apply custom colors to sub-menus/sub-items.

Tasks on this PR:

- [x] Propagate the color changes for background and text through to all submenus


## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->
![image](https://user-images.githubusercontent.com/77539/73002353-a9dda580-3de2-11ea-83e0-c64b2c75cb37.png)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

retrofox committed:
navigation-link: add has-child css class

retrofox committed:
colors: move navigation scss var to base-styles

retrofox committed:
squash

retrofox committed:
navigation: organize

retrofox committed:
navigation: propagate style css Class

retrofox committed:
navigation: implement dark styles

frontdevde committed:
Minor wip style update to be closer to design

retrofox committed:
navigation: remove duplicated has-child code

retrofox committed:
navigation-link: remove unused prop

apeatling commented on packages/block-library/src/navigation/editor.scss:
```diff
@@ -77,14 +77,62 @@ $navigation-item-height: 46px;
+		background-color: $dark-style-sub-menu-background-color;
+	}
+}
+
```
Is these styles going to be portable to the front end and shared?

retrofox replied:
> Are these styles going to be portable to the front end and shared?

Probably(?). I left the code there in case it doesn't set up background color. Guessing that transparent could be a problem. So it will apply the default background-color as long it doesn't already define a custom one.
Something to keep discussing, btw.

retrofox replied:
Something like this:
```css
	// No background color - Default / Light styles
	&:not(.has-background-color) .wp-block-navigation-link > .block-editor-inner-blocks {
		background-color: $light-style-sub-menu-background-color;
	}
```

apeatling replied:
I was talking about all the styles, not just the lines above. I was hoping we'd have a single stylesheet that can apply navigation structure to both the editor and the front end. Colors could be separated so they apply even in situations where you've set a different block style in the future and the structural look is totally different.

retrofox replied:
>  ... was hoping we'd have a single stylesheet that can apply navigation structure to both the editor and the front end.

That makes totally sense to me. Let's find a way to isolate and consume these from both sides.

retrofox committed:
navigation: remove styles - propagate custom colors.

retrofox committed:
navigation: fix has bgc css class name

retrofox committed:
navigation: apply light/dark styles when no colors

retrofox commented on packages/block-library/src/navigation-link/editor.scss:
```diff
@@ -66,6 +66,30 @@
+			position: absolute;
+			right: 0;
+			top: 50%;
+			border-left: 4px solid transparent;
```
Can we consider using an icon or char instead of a border technique? Keep in mind that, at least for now, we have only two ways to apply colors: `color` and `background-color`. Applying border-color will be a pain.
We can use dashicons or even SVG.

retrofox committed:
navigation-link: propagate Navigation block attrs

retrofox committed:
navigation-link: add border-color inline style

retrofox committed:
navigation-link: apply custom border-color

retrofox committed:
navigation-link: set colors CSS classes

retrofox committed:
use-colors: add attr for color value

retrofox committed:
navigation: store and populate inline colors
It stores the color values for both text and background color in the Navigation block, and propagete these values to the Navigation Link block, in order to be able to apply colors using inline style.

retrofox committed:
navigation: improve handling colors/styles.

frontdevde committed:
remove old submenu frontend styles

retrofox committed:
navigation: fix env linting error

retrofox committed:
add wip version of children arrow indicators
Need to define what design keep: CSS arrow vs SVG icon

retrofox committed:
navigation: fix rebasing issues

retrofox committed:
navigation-link: remove CSS arrow in favor of SVG

retrofox committed:
navigation: detect and propagate color to the SVG

retrofox committed:
navigation-item: use detected color as default

retrofox committed:
navigation-link: add chevron to all menu levels

frontdevde committed:
Show submenu indicators by default

frontdevde committed:
Fix linting and formatting errors

frontdevde committed:
Update Display Settings copy to reflect that it now affects multiple levels

frontdevde committed:
Update submenu indicators to new design

frontdevde committed:
Remove obsolete indicator styles

frontdevde committed:
Start cleanup of style.scss by ordering by function of styles

frontdevde committed:
Reduce specificity by not overwriting inline styles

frontdevde committed:
Continue organizing style.scss by ordering by function of styles

frontdevde committed:
Share basic top level styles between editor and frontend

frontdevde committed:
readd rotation of submenu indicators

retrofox commented on packages/block-library/src/navigation-link/editor.scss:
```diff
@@ -32,11 +32,11 @@
 	.wp-block-navigation-link__content {
 		display: flex;
 		align-items: center;
+		width: max-content;
```
💯

frontdevde committed:
make child class usage consistent between editor and frontend

frontdevde committed:
add class to frontend to apply styles that match .block-editor-inner-blocks usage

frontdevde committed:
start sharing editor styles with frontend

retrofox committed:
navigation-link: apply svg text color with CSS

frontdevde committed:
update navigation container positioning

frontdevde committed:
Add submenu indicators to submenus on the frontend

frontdevde committed:
unify submenu icon styles

frontdevde committed:
move icon rotation to correct position

frontdevde committed:
apply classes to all navigation links

frontdevde committed:
add colors to frontend submenus

retrofox committed:
navigation-link: put the icon on the full right

retrofox committed:
navigation: remove unneeded padding

retrofox committed:
navigation: remove deprecated styles

retrofox committed:
navigation: hid all sub-menus

retrofox committed:
navigation: remove detected-color code

retrofox committed:
navigation: unify default styles.

retrofox committed:
navigation: dealing with min heigh and vert align

retrofox committed:
navigation: main menu height / vertical alignment

retrofox commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -77,11 +77,18 @@ $navigation-height: 60px;
+			min-height: $navigation-min-height;
+			margin-top: 0;
+			margin-bottom: 0;
+		}
```
@frontdevde, not sure if it's the correct implementation. The idea behind this is to fit the `.wp-block-navigation-link` in height to its container.

<img src="https://user-images.githubusercontent.com/77539/72985435-bb17b980-3dc4-11ea-91b3-6e6e6535a764.png" width="500px" />

retrofox committed:
navigation: adjust box model and position

retrofox committed:
navigation: set width fixing the alignment

retrofox committed:
navigation: set CSS classes depending on value att

obenland commented:
@shaunandrews Do you think you'll have a chance today to try out these changes?

apeatling commented:
This is starting to come together nicely! I think it could use a design lookover, especially around padding, borders, and spacing both in the editor and front end.

obenland commented on packages/block-library/src/navigation/index.php:
```diff
@@ -271,12 +267,18 @@ function register_block_core_navigation() {
 				'customTextColor'       => array(
 					'type' => 'string',
 				),
+				'valueTextColor'        => array(
```
Which "value" is that? Could we find a more descriptive name here?

retrofox replied:
This value (already removed) came from the use-colors hook. I had modified introduced the ability to set up this attribute from there, too.
The reason was to be able to pick up the color value of the text, independently if it was defined using a custom color or a palette color, to apply to the border-color property.
I've removed all this stuff and replaced using the CSS currentColor property.

retrofox replied:
I've restored these changes. Going to explain why in the use-colors hook change.

retrofox replied:
Basically this attribute will store the color value beyond how it was defined, either picking up from the colors palette or using the custom color panel.

obenland committed:
Slightly simplify css class assembly

obenland commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -147,12 +148,23 @@ function NavigationLinkEdit( {
+						'is-selected': isSelected,
+						'has-link': !! url,
+						'has-child': hasDescendants,
+						'has-text-color': navigationBlockAttributes.valueTextColor,
```
Can `navigationBlockAttributes` be destructured to make this a little easier to grok? Or are these variable names already in use?

retrofox replied:
Agree. let me rename or destructure it

retrofox committed:
navigation: fix and simlify block styles

retrofox committed:
navigation: set border color using currentColor

retrofox committed:
navigation-edit: simplify color attrs

obenland commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -1,236 +1,269 @@
+	}
+
+	// Main menu
+	.wp-block-navigation-link {
```
Should this be in the navigation-link component?

retrofox replied:
We've been talking about where we should put this code. Although it clearly belongs to the navigation-link component, splitting out the code of the sub-menus structure in two files will make understanding the code otherwise tough labor. 
So we agreed to keep it in the same file, organized by sections. We also considered creating a new file only to implement the menu structure, but it's unneeded at least for now.

obenland commented:
Those front-end paddings seem to be uneven and have a much higher value than in the editor

![image](https://user-images.githubusercontent.com/1398304/73021222-038ea180-3ddc-11ea-9330-0367b9c29181.png)

retrofox commented on packages/block-editor/src/components/colors/use-colors.js:
```diff
@@ -194,6 +194,7 @@ export default function __experimentalUseColors(
 						[ color ? name : camelCase( `custom ${ name }` ) ]: color ?
 							color.slug :
 							newColor,
+						[ camelCase( `value ${ name }` ) ]: newColor,
```
This line will set the `value${ name }` attribute. It allows getting the color value in the `rgb()` notation, very useful when we need to apply them using inline styles beyond how the color has been defined (through the colors palette, custom color).
It's already handled by the hook, populating the `style` attribute of its children component, for instance...

```jsx
<TextColor>
    <MyComponent />
</TextColor>
```

The style attribute of the `<MyComponent />` will get populated by the parent `<TextColor />` one.

The problem is, on this implementation, in particular, we need to know the `rgb()` value for the inner blocks (NavigationItem). We need the ability to get these values, store them in the Navigation attributes, to finally collect them from the `<NavigationItem />` using the `getBlockAttributes()` store selector.

Pinging to @epiqueras to get a qualified opinion about this change.

epiqueras replied:
You can just use `TextColor.color`.

retrofox replied:
> You can just use `TextColor.color`.

Yup, I already know that it's possible to pick up the property from the component.
Anyway, we need to store these values into the attributes object since it's needed to pull them from the inner blocks.
Thanks for your advice!

retrofox committed:
navigation: improve paddings, positions, etc.

retrofox committed:
navigation: pick up rgb colors outside of hook

retrofox committed:
navigation: move sub-menus (L1) up a grid size

retrofox committed:
navigation: fix php linting errors

retrofox commented:
> This is starting to come together nicely! I think it could use a design lookover, especially around padding, borders, and spacing both in the editor and front end.

> Those front-end paddings seem to be uneven and have a much higher value than in the editor
> 
> ![image](https://user-images.githubusercontent.com/1398304/73021222-038ea180-3ddc-11ea-9330-0367b9c29181.png)

I've been working on the box model for the navigation Items. After the last changes, it applies almost the same values to both sides editor-canvas and front-end.

*front-end*
<img src="https://user-images.githubusercontent.com/77539/73035005-6fe1c300-3e25-11ea-9887-d6d15f59a2ae.png" width="600px" />

*editor-canvas*
<img src="https://user-images.githubusercontent.com/77539/73035016-82f49300-3e25-11ea-9c60-7a316a4111c2.png" width="600px" />

All these values are defined using the `$grid-size` as the measuring unit.

shaunandrews commented:
I took this for a spin this morning. Here's a few things I found:

<img width="478" alt="image" src="https://user-images.githubusercontent.com/191598/73080170-668f3f80-3e93-11ea-942e-fd8cb56cea93.png">

In Safari the dropdown indicator on top-level items is pointing the wrong way. Also, there are two insert (+) icons. Can we try hiding the top-level inserter icon when a submenu has focus?

--

<img width="303" alt="image" src="https://user-images.githubusercontent.com/191598/73080805-6e031880-3e94-11ea-86b2-74dd0277afc2.png">

The insert (+) icon in the submenu needs some whitespace below it; Its current hitting the bottom border of the submenu.

--

<img width="271" alt="image" src="https://user-images.githubusercontent.com/191598/73080307-9e968280-3e93-11ea-8f1c-b02bf37982b0.png">

I'm able to trigger the display of the "in-between" inserter. I think we need to disable this.

--

<img width="349" alt="image" src="https://user-images.githubusercontent.com/191598/73080375-bd951480-3e93-11ea-9153-e227a93df69f.png">

The top-level inserter is misaligned. Maybe this is a theme thing? I'm using TwentyTwenty.

--

<img width="548" alt="image" src="https://user-images.githubusercontent.com/191598/73080452-d7cef280-3e93-11ea-9b4d-5c935fa54f35.png">

When previewing, the arrow is pointing in the wrong direction. Also, the submenu is getting cut off as the page is otherwise empty. I'm not sure what we could do about this, as the theme is setting `overflow: hidden;` on the `#site-content` div.

--

Could we try showing the submenus when hovering the block (along when the parent block is selected), mimicking the behavior on the frontend? It might feel overwhelming, but I keep hovering a parent item expecting the submenu to show.

--

I don't _love_ the way the submenus adopt the color settings. I'd love to expose options for setting colors on the top-level _and_ submenus separately. However, I don't wanna just tack on more options like this:

<img width="540" alt="image" src="https://user-images.githubusercontent.com/191598/73081163-1d3fef80-3e95-11ea-899d-6c66feb967e0.png">

For now, I think the current behavior is ok. There's the beginning of some work around redesigning the color picker (#19785) — that, along with #19344 could make the UI for more color options better.

obenland commented:
Thanks so much for your feedback @shaunandrews! 

In the interest of unblocking other design issues, can we focus on regressions and fix the other things in separate PRs?

These are the things that look like regressions to me:
* The insert (+) icon in the submenu needs some whitespace below it
* The top-level inserter is misaligned. Maybe this is a theme thing? I'm using TwentyTwenty.
* When previewing, the arrow is pointing in the wrong direction.

One note that might warrant some more discussion:
* Could we try showing the submenus when hovering the block

apeatling committed:
Update the inserter positioning and spacing. Remove the submenu borders so they belnd with the top level menu.

apeatling commented:
I could not reproduce the menu cutoff or incorrectly pointing arrow on the front end. I've fixed the inserter spacing, the top level inserter misalignment, and also removed the submenu borders so that colors flow together more and it looks more like a single menu:

<img width="780" alt="Screen Shot 2020-01-24 at 11 23 01 AM" src="https://user-images.githubusercontent.com/1464705/73097562-5039b180-3e9c-11ea-8598-f54fc4d82348.png">

apeatling commented:
> Could we try showing the submenus when hovering the block (along when the parent block is selected), mimicking the behavior on the frontend? It might feel overwhelming, but I keep hovering a parent item expecting the submenu to show.

I think this is going to be problematic, purely because submenus are going to be popping open as you move the mouse around the editor doing other things. To do this we would need to add some sort of hover intent so that you need to pause first on a menu. I think we can consider it, but not as part of this PR.

apeatling committed:
Restore the border on sub menus, but use alpha so that it blends better with whatever the selected background color is.

jeryj committed:
Align flyout submenu tops.

jeryj committed:
Use $border-width instead of a magic number for submenu alignment

jeryj committed:
Removed ul and li padding on submenus and moved them into the link <a> level so links are now full width and everywhere in the menus is hoverable/clickable. Focus rings look nicer now too.

jeryj committed:
Set a fixed width on submenu widths instead of relying on the content width.

This means that long submenu items will wrap. I also set a lower line height on items so that it was more obvious that a wrapped submenu item was one item,

jeryj committed:
Move padding from li to a on top-level items.

jeryj committed:
Moved caret on submenu icon left a bit more.

obenland committed:
Better align grandchild links with bottom child links

obenland reviewed (approved):
@retrofox @frontdevde Feel free to do a final round of reviews, but I think this is in good enough shape to get in now

apeatling commented:
@jeryj I noticed we've lost the correct amount of top padding on sub menu items:

<img width="494" alt="Screen Shot 2020-01-28 at 9 31 31 AM" src="https://user-images.githubusercontent.com/1464705/73289290-53d97b00-41b1-11ea-9d37-167a3ef4659e.png">

Compare that to my screenshot above, the top and bottom padding is the same.

retrofox committed:
navigation: refact -> icons as components

retrofox committed:
navigation: remove is level zero arg

retrofox committed:
navigation: remove flex-wrap in to fix nav width

retrofox committed:
navigation: handling default true value for arrow

jeryj committed:
Set top: 100% on navigation submenu instead of magic number

jeryj committed:
Left align submenus

retrofox committed:
navigation: fix showing submenu icon

retrofox committed:
navigation: fixing linting errors

retrofox committed:
Merge branch 'update/navigation-submenu-new-design' of github.com:WordPress/gutenberg into update/navigation-submenu-new-design

retrofox committed:
navigation: fix setting link width

retrofox commented:
Thanks, folks for reviewing and helping with the development. 👍
